### PR TITLE
NPoS Challenge Mode

### PR DESCRIPTION
**Pull request by coriolinus**
Builds on #8160.

- Updates the PJR check to return a counterexample if one exists
- Adds functions to cheaply check counterexamples

This is in support of off-chain PJR challenges: if a miner discovers
that an accepted election solution does not satisfy PJR, it will be
eligible for substantial rewards. This helps ensure that validator
elections have an absolute quality floor, so even if someone manages
to censor well-behaved solutions to give themselves unfair representation,
we can catch them in the act and penalize them.

### CODE REVIEW

**Comment by kianenigma** [comment]
Screenshot 2021-03-02 at 06 40 13


You probably intended to change this to merge into your other branch?

**Comment by coriolinus** [comment]
I'll follow whatever the convention is; my intent was to rebase once #8160 is merged.

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:330]:
```suggestion
	counter_example: AccountId,
```

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:312]: nit: I think the order of functions in this file is not logical. I'd say it should be high -> low abstraction. So Iexpect to first see the functions that auto-compute the `t`. Then, the functions that accept `t`. Then, `_core` stuff, and finally `slack` and `prescore`. 

Would be nice if you fix this in the other PR as well.

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:355]: I haven't read the whole code, but this is wrong IMO. The whole point of the challenge is to _not_ need to go over _all candidate_, and the challenge would just say which candidate we must look into exactly. Thus, you should not need to pass in all the candidates again. 

Voters are sadly unavoidable at first glance, but I guess we can improve that as well.

**Review by kianenigma — CHANGES_REQUESTED** [review]
API is confused. The challenge, ideally has just one input: 

1. `counter_example` candidate. Next to this, we will need to pass in all the voters as well, and nothing more to validate the challenge. 

From this, we just compute `prescore(counter_example, voters, t)`. 

So no need to pass `support` and all_candidates` in.

**Comment by coriolinus** [comment]
The code doesn't iterate over all candidates, except to the degree that it searches the candidate list for the one of interest. However, it does require the full list of supports, all candidate, all voters. The chain of requirements goes like this:

- To validate a PJR challenge fundamentally requires calculating its `pre_score`
- `pre_score` requires calculating the `slack` for each voter
- `slack` requires an accurate list of the voter's edges: all the candidates it voted for, and by how much
- Therefore we need, bare minimum, `&[Voter<AccountId>]`. 
- `Voter` is a struct with private fields and a complicated internal structure. It would be hard for users to accurately construct it on their own even if we made the fields public or provided a constructor.
- We can construct the list of voters simply by reusing `fn prepare_pjr_input`, which requires the full list of supports, candidates, voters.

A reasonable argument could be, "but why not extract a subset of `prepare_pjr_input` which can handle constructing the inputs relevant to a single candidate?" My answer to that argument is that it's more trouble than it's worth. This module assumes that the lists of supports, all candidates, all voters are available to modules who care about the election. Making those inputs a prerequisite for counterexample proofs seems to be a small additional burden.

File: primitives/npos-elections/src/pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs:355]: I understand the motivation, but I think we can't avoid requiring those inputs. The elections code requires several complicated data structures: `Candidate`, `Voter`, `CandidatePtr`, which inter-relate to each other in complicated ways. If those complicated data structures were universal, then we could just have the calling code pass them in. As things stand, we have to construct them. 

"Why not add have the calling module cache the `Vec<Voter<AccountId>>` and restore it, instead of reconstructing it here?" Because `Voter` contains several `Edge`s, and `Edge` contains a `CandidatePtr` which is a `Rc<RefCell<Candidate<AccountId>>>>`, and ref-counted pointers are a pain to serialize and deserialize properly.

**Comment by kianenigma** [comment]
> My answer to that argument is that it's more trouble than it's worth. This module assumes that the lists of supports, all candidates, all voters are available to modules who care about the election. Making those inputs a prerequisite for counterexample proofs seems to be a small additional burden.

We will surely need to run the `pjr-counter-example` check on-chain and per proof that we receive. I suppose this could be due to you not being yet very familiar with the internals of frame, but something that runs on-chain and can be triggered by the user is def. worth being optimized. 

All in all, we could also meet in the middle and merge this as-is and update the API once the challenge phase is in. But note that that is not imminent and we first need to finish the unsigned + signed phase, so I see it happening in the next ~3 months or so, based on how fast we move.

File: primitives/npos-elections/src/pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: 030dc6051

File: primitives/npos-elections/src/pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: b36f71895 and c1f1ca718

**Comment by coriolinus** [comment]
> but something that runs on-chain and can be triggered by the user is def. worth being optimized.

Could we apply an economic incentive instead? We're already planning on incentivizing users to submit counterexamples by providing some reward to them. Couldn't we also apply some penalty in the even that the user submits an invalid counterexample?

**Comment by coriolinus** [comment]
Reduced to unique commits; should be easier to review now.

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:154]:
```suggestion
		Some((max_pre_score, counter_example)) if max_pre_score >= t => Err(counter_example),
		_ => Ok(()),
	}
```

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:171]: One random note unrelated to this PR per se: this pallet had quite a bad rap for not having a well defined behaviour with duplicate votes and candidates and targets etc. Now, finally, everything that goes through `setup_input` (called prior to `seq_phragmen` and `phragmms`) gets sanitized. Would be good to ensure these functions have the same standard.

**Review by kianenigma — APPROVED** [review]
The only thing I see missing is a test. Otherwise looks clean.

File: primitives/npos-elections/src/pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs:167]: [`prepare_pjr_input`](https://github.com/paritytech/substrate/blob/2d695630e3eabfa9b0b2be8d8e19f16dfe70e323/primitives/npos-elections/src/pjr.rs#L230-L261) mentions `setup_inputs` in its documentation, but doesn't actually delegate. Shall I create an issue to add explicit delegation there to ensure that the behaviors are identical?

File: primitives/npos-elections/src/pjr.rs
Comment (shawntabrizi) [inline primitives/npos-elections/src/pjr.rs:205]: this exact text is copied on all these functions?

**Review by shawntabrizi — APPROVED** [review]
looks clean

File: primitives/npos-elections/src/pjr.rs
Comment (shawntabrizi) [reply primitives/npos-elections/src/pjr.rs:205]: Also any advantage to breaking up the steps into such minimal functions? I feel if anything, this may actually loose context of what other operations are happening between the steps, where optimizations may be added.

Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs:205]: The difference in the text is 

```diff
- /// This function uses the standard threshold.
+ /// This function uses a supplied threshold.
```

The documentation goal is to ensure that each function's documentation completely describes what that function does, even though a similar function exists.

As for why the functions are factored in this way, it comes down to two things. On the one hand, it mirrors the factorization of the `pjr_check` functions. On the other hand, it means that users can supply a threshold, or not, as desired. We have use cases for both.

**Comment by coriolinus** [comment]
bot merge

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [reply primitives/npos-elections/src/pjr.rs:167]: An issue would be good, bit I am not sure if delegation is the best way, we have to see. We can maybe mix this into #4593

**Pull request merged**
